cozy/cozy-flagship-app
# feat: implement sharing files navigation (android + iOS)

### Data flow:
```mermaid
graph TD
    subgraph State Layer
        SProvider[Sharing Provider]
    end

    subgraph View Layer
        SS[SharingScreen]
    end

    subgraph User Flow Layer
        UGA[useGlobalAppState]
    end

    subgraph Business Layer
        SH[SharingStatus]
        SService[SharingData]
    end

    subgraph Native Layer
        SIM[SharingIntentModule\nAndroid]
        RNRSI[RNReceiveSharingIntent]
    end

    SH -->|"SharingIntentStatus\nenum"| SProvider
    SService -->|"filesToUpload\nReceivedFile[]"| SProvider
    SProvider -->|"filesToUpload\nReceivedFile[]"| SS
    SProvider -->|"sharingIntentStatus\nenum"| UGA
    SIM -->|"wasOpenedViaSharing()\nboolean"| SH
    RNRSI -->|"Linking.openUrl()\niOS"| SH
    RNRSI -->|"ReceiveSharingIntent\ngetReceivedFiles()"| SService

    style SS fill:slateblue
    style UGA fill:darkslateblue
    style SH fill:darkslategray
    style SService fill:darkslategray
    style SIM fill:indianred
    style SProvider fill:forestgreen
    style RNRSI fill:indianred
```

### User flow:
```mermaid
graph TD
    A[User Opens App] --> |Without File Sharing| B
    A --> |With File Sharing| E
    B --> |No Lock Screen| C[Home Screen]
    B --> |Has Lock Screen| D
    D --> |Unlocks Device| C
    D --> |Stuck| D1[Stuck on Lock Screen]
    E --> |No Lock Screen| F[File Sharing Screen]
    E --> |Has Lock Screen| G
    F --> H[Home Screen After Upload Confirmation]
    G --> |Unlocks Device| F
    G --> |Stuck| G1[Stuck on Lock Screen]
```

### End to end manual testing 

- [x] **Cold Start with Sharing Intent**:
    - Ensure the app is completely closed (not in the background).
    - Share content to your app from another app.
    - Verify that the app opens and detects the sharing intent (`sharingIntentStatus` should be `undetermined` then `openedViaSharing`).

- [x] **Cold Start without Sharing Intent**:
    - Ensure the app is completely closed.
    - Open the app normally (without sharing).
    - Verify that the app opens and detects the sharing intent (`sharingIntentStatus` should be `undetermined` then `notOpenedViaSharing`).

- [x] **Resume from Background with Sharing Intent**:
    - Open the app normally.
    - Put the app in the background.
    - Share content to your app from another app.
    - Verify that the app comes to the foreground and detects the sharing intent (`sharingIntentStatus` should be `openedViaSharing`).

- [x] **Resume from Background without Sharing Intent**:
    - Open the app normally.
    - Put the app in the background.
    - Bring the app to the foreground without a sharing intent (e.g., from the recent apps list).
    - Verify that the app does not detect a sharing intent (`sharingIntentStatus` should be `notOpenedViaSharing`).

- [x] **Switching Intents**:
    - Open the app with a sharing intent.
    - Put the app in the background.
    - Bring the app to the foreground without a sharing intent.
    - Verify that the app does not detect a sharing intent (`sharingIntentStatus` should be `notOpenedViaSharing`).
    - Put the app back in the background.
    - Share content to your app from another app.
    - Verify that the app detects the sharing intent (`sharingIntentStatus` should be `openedViaSharing`).

- [x] **App Restart**:
    - Open the app with a sharing intent.
    - Restart the app (either via a setting in the app or by force-closing and reopening).
    - Verify that the app does not detect a previous sharing intent (`sharingIntentStatus` should be `notOpenedViaSharing`).
